### ISSUE TITLE
"cross compiling" to native (build) machine leads to different install paths for libraries

### ISSUE THREAD

**Issue by baryluk** [mesonbuild/meson#6115]
I am using `meson 0.52.0` on Debian Linux testing amd64.

Set some build machine stuff:

```shell
export CC="/usr/bin/x86_64-linux-gnu-gcc-${GCCVERSION}"
export CXX="/usr/bin/x86_64-linux-gnu-g++-${GCCVERSION}"
export PKG_CONFIG="/usr/bin/x86_64-linux-gnu-pkg-config"
```

Prepare cross-file:

```shell
cat > "${MESONCROSSFILE_64}" <<EOF
[binaries]
c = '/usr/bin/x86_64-linux-gnu-gcc-${GCCVERSION}'
cpp = '/usr/bin/x86_64-linux-gnu-g++-${GCCVERSION}'
llvm-config = '/usr/bin/llvm-config-${LLVMVERSION}'
strip = '/usr/bin/x86_64-linux-gnu-strip'
EOF
```

Do a build and install without cross-file:

```shell
cd "${DRMREPO}"
rm -rf "${DRMBUILDDIR}/build-amd64-opt/"
meson "${DRMBUILDDIR}/build-amd64-opt/" \
  --buildtype=plain \
  --prefix="${INSTALLDIR}/build-amd64-opt/install"
ninja -C "${DRMBUILDDIR}/build-amd64-opt/"
ninja -C "${DRMBUILDDIR}/build-amd64-opt/" install
```

Relevant output:
```
...
Build type: native build
Project name: libdrm
Project version: 2.4.100
C compiler for the host machine: /usr/bin/x86_64-linux-gnu-gcc-9 (gcc 9.2.1 "x86_64-linux-gnu-gcc-9 (Debian 9.2.1-15) 9.2.1 20191027")
C linker for the host machine: GNU ld.bfd 2.33.1
Host machine cpu family: x86_64
Host machine cpu: x86_64
Run-time dependency threads found: YES 
Found pkg-config: /usr/bin/x86_64-linux-gnu-pkg-config (0.29)
...
Installing intel/libdrm_intel.so.1.0.0
  to /home/user/mesa-git/installdir/build-amd64-opt/install/lib/x86_64-linux-gnu
...
Installing /home/user/mesa-drm-git/builddir/build-amd64-opt/meson-private/libdrm_amdgpu.pc
  to /home/user/mesa-git/installdir/build-amd64-opt/install/lib/x86_64-linux-gnu/pkgconfig
...
```

Do a build and install with cross-file:

```shell
cd "${DRMREPO}"
rm -rf "${DRMBUILDDIR}/build-amd64-opt/"
meson "${DRMBUILDDIR}/build-amd64-opt/" \
  --buildtype=plain \
  --cross-file="${MESONCROSSFILE_64}" \
  --prefix="${INSTALLDIR}/build-amd64-opt/install"
ninja -C "${DRMBUILDDIR}/build-amd64-opt/"
ninja -C "${DRMBUILDDIR}/build-amd64-opt/" install
```

Relevant output:
```
Build type: cross build
...
C compiler for the build machine: /usr/bin/x86_64-linux-gnu-gcc-9 (gcc 9.2.1 "x86_64-linux-gnu-gcc-9 (Debian 9.2.1-15) 9.2.1 20191027")
C linker for the build machine: GNU ld.bfd 2.33.1
WARNING: Env var CC seems to point to the cross compiler.
This is probably wrong, it should always point to the native compiler.
C compiler for the host machine: /usr/bin/x86_64-linux-gnu-gcc-9 (gcc 9.2.1 "x86_64-linux-gnu-gcc-9 (Debian 9.2.1-15) 9.2.1 20191027")
C linker for the host machine: GNU ld.bfd 2.33.1
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: x86_64
Host machine cpu: x86_64
Target machine cpu family: x86_64
Target machine cpu: x86_64
Run-time dependency threads found: YES 
Found pkg-config: /usr/bin/pkg-config (0.29)
...
...
Installing intel/libdrm_intel.so.1.0.0
  to /home/user/mesa-git/installdir/build-amd64-opt/install/lib
...
Installing /home/user/mesa-drm-git/builddir/build-amd64-opt/meson-private/libdrm_amdgpu.pc
  to /home/user/mesa-git/installdir/build-amd64-opt/install/lib/pkgconfig
```

How do I make meson make always install into `lib/triplet/` and `lib/triplet/pkgconfig`, even when it detects build as cross build? I tried setting `--buildtype=plain` but it  doesn't help. Meson still installs with no triplets in install paths.

The same happens when cross compiling on amd64 to i386. Installed paths don't have triplets in them.

**Comment by baryluk** [comment]
I managed to make it work by putting:

```
[paths]
libdir = 'lib/x86_64-linux-gnu'
```

In cross-file, but why default is different for native vs cross build? And how meson decides to use lib/triplet when doing native build, and where it gets the triplet from?
 `/usr/bin/x86_64-linux-gnu-gcc-9 -dumpmachine` ?

**Comment by jpakkane** [comment]
When compiling natively it is easy to detect what the libdir should be (see `default_libdir` in `mesonlib.py`) . This is not the case when cross compiling. In general it is impossible to know how your cross target has set its root file system. We use `lib` as the default in this case because that is the safe default. The correct fix, as you found, is to explicitly define the libdir.

**Comment by dabrain34** [comment]
in meson master branch 200738a3e6b48671aac2865c304dded96179e4ac, it seems that a cross build uses 'lib/' + archpath for default_libdir  which leads to `lib/x86_64-linux-gnu` on my debian like system for an arm target. Is it the expected behavior ? 

I mean it looks to me a regression  and should be only `lib` in a cross build configuration. Is it correct ?

**Comment by nirbheek** [comment]
Yes, we should always default to `lib` when cross-compiling.

**Comment by dabrain34** [comment]
i confirm a regression ...

**Comment by dabrain34** [comment]
Here is the guilty commit 3ba0073df6014d4b594df052f8e1fa75ce17b20e

**Comment by dabrain34** [comment]
https://github.com/mesonbuild/meson/pull/6718

**Comment by Jehan** [comment]
Hi!

Since yesterday, our cross-builds on GIMP CI got broken. Looking a bit at our build logs, I discover it now installs our self-built dependencies to `libdir=${prefix}/lib/x86_64-linux-gnu` for our Windows builds, which is obviously wrong (as confirmed by above comments). I assume the bug came with meson 0.54.1 as the inclusion to Debian/testing (our base system for CI) is yesterday.

And this is apparently the bug the last comments here talk about but I am a bit unclear whether this is fixed or not. I see a PR #6718, merged then reverted by #7076. But the Debian-side report from 4 days seems to say that it is all good now in meson 0.54.1 in Debian package. It's not.

For the record, we had a cross file, but no libdir set (neither through command line nor inside the cross file). Until now we were just relying on meson installing correctly in `$prefix/lib` (which it was doing fine until 2 days ago). We are going to set an explicit libdir for the time being, but it would be nice if it could be fixed.

**Comment by jpakkane** [comment]
Can you test if the current master works (it should, there is a test even)? That would tell whether the problem is with the Debian patch or something else.

**Comment by baryluk** [comment]
> Yes, we should always default to `lib` when cross-compiling.

Well, that is not what I want. How I get a reverse behavior? To always default to `lib/<triplet>`  instead.

**Comment by jpakkane** [comment]
It's not really possible for Meson to know when cross compiling what the output libdir name should be (you could be cross compiling to a different Linux distro that names things differently, for example). Thus we default to `lib`, which is the safe choice. You can override this [in your cross file](https://mesonbuild.com/Machine-files.html#paths-and-directories).

**Comment by Jehan** [comment]
> Can you test if the current master works (it should, there is a test even)? That would tell whether the problem is with the Debian patch or something else.

Testing with meson master code on a Debian/testing CI, with a cross file and no libdir set:

```
 Cloning into 'babl'...
The Meson build system
Version: 0.54.999
Source dir: /builds/GNOME/gimp/_deps/babl
Build dir: /builds/GNOME/gimp/_deps/babl/_build
Build type: cross build
Project name: babl
Project version: 0.1.75
C compiler for the build machine: cc (gcc 9.3.0 "cc (Debian 9.3.0-11) 9.3.0")
C linker for the build machine: cc ld.bfd 2.34
C compiler for the host machine: x86_64-w64-mingw32-gcc (gcc 9.3.0 "x86_64-w64-mingw32-gcc (GCC) 9.3-win32 20200324")
C linker for the host machine: x86_64-w64-mingw32-gcc ld.bfd 2.34
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: x86_64
Host machine cpu: x86_64
Target machine cpu family: x86_64
Target machine cpu: x86_64
Message: Host machine cpu family: x86_64
Message: Host os: windows
Message: Build os: linux
[…]
 Installing /builds/GNOME/gimp/_deps/babl/_build/meson-private/babl.pc to /builds/GNOME/gimp/.local/share/crossroad/roads/w64/gimp/lib/x86_64-linux-gnu/pkgconfig
```

So yes, unfortunately it happens with bare meson code without Debian patches.

**Comment by jpakkane** [comment]
Hmmmmmmmm. This is probably a different bug with the same outcome then...

**Comment by nirbheek** [comment]
I can confirm that https://github.com/mesonbuild/meson/pull/7076 breaks it. This works fine in 0.54.1.

**Comment by nirbheek** [comment]
The test is faulty because we pass `--libdir` unconditionally.

**Comment by dabrain34** [comment]
Yes I confirm too that the revert breaks the cross build to arm originally fixed with #6718

I can try to fix the issue where --libdir is not taken into account.

**Comment by dabrain34** [comment]
I gave it a try, reverting the revert #7076 ( so having the initial bug fix) and I dont have any issue with zlib:

First I added a log in msetup.py
```
diff --git a/mesonbuild/msetup.py b/mesonbuild/msetup.py
index 252151194..eb94718f8 100644
--- a/mesonbuild/msetup.py
+++ b/mesonbuild/msetup.py
@@ -184,6 +184,7 @@ class MesonApp:
         mlog.log('Host machine cpu:', mlog.bold(intr.builtin['host_machine'].cpu_method([], {})))
         logger_fun('Target machine cpu family:', mlog.bold(intr.builtin['target_machine'].cpu_family_method([], {})))
         logger_fun('Target machine cpu:', mlog.bold(intr.builtin['target_machine'].cpu_method([], {})))
+        mlog.log('libdir:', mlog.bold(env.get_libdir()))
         try:
             if self.options.profile:
                 fname = os.path.join(self.build_dir, 'meson-private', 'profile-interpreter.log')
```

Then here is some tests with zlib:
```
$ MESON build

The Meson build system
Version: 0.54.999
Source dir: /tmp/zlib
Build dir: /tmp/zlib/build
Build type: native build
Project name: zlib
Project version: 1.2.11
C compiler for the host machine: ccache cc (gcc 7.5.0 "cc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
C linker for the host machine: cc ld.bfd 2.30
Host machine cpu family: x86_64
Host machine cpu: x86_64
libdir: lib/x86_64-linux-gnu
Build targets in project: 1

Found ninja-1.8.2 at /usr/bin/ninja
```

```
$ MESON build --cross-file=my-cross-file.txt

The Meson build system
Version: 0.54.999
Source dir: /tmp/zlib
Build dir: /tmp/zlib/build
Build type: cross build
Project name: zlib
Project version: 1.2.11
C compiler for the build machine: ccache cc (gcc 7.5.0 "cc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
C linker for the build machine: cc ld.bfd 2.30
C compiler for the host machine: aarch64-linux-gnu-gcc (gcc 7.5.0 "aarch64-linux-gnu-gcc (Ubuntu/Linaro 7.5.0-3ubuntu1~18.04) 7.5.0")
C linker for the host machine: aarch64-linux-gnu-gcc ld.bfd 2.30
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: aarch64
Host machine cpu: arm64
Target machine cpu family: aarch64
Target machine cpu: arm64
libdir: lib
Build targets in project: 1

Found ninja-1.8.2 at /usr/bin/ninja
```
```
$ MESON build --cross-file=my-cross-file.txt --libdir=lib/arm

The Meson build system
Version: 0.54.999
Source dir: /tmp/zlib
Build dir: /tmp/zlib/build
Build type: cross build
Project name: zlib
Project version: 1.2.11
C compiler for the build machine: ccache cc (gcc 7.5.0 "cc (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0")
C linker for the build machine: cc ld.bfd 2.30
C compiler for the host machine: aarch64-linux-gnu-gcc (gcc 7.5.0 "aarch64-linux-gnu-gcc (Ubuntu/Linaro 7.5.0-3ubuntu1~18.04) 7.5.0")
C linker for the host machine: aarch64-linux-gnu-gcc ld.bfd 2.30
Build machine cpu family: x86_64
Build machine cpu: x86_64
Host machine cpu family: aarch64
Host machine cpu: arm64
Target machine cpu family: aarch64
Target machine cpu: arm64
libdir: lib/arm
Build targets in project: 1

Found ninja-1.8.2 at /usr/bin/ninja
```
